// teacherControl/index.js
var mongo = require("mongodb");

function getOtherScore(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.find({}).toArray(function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            var d = {};
                            d.data = [];
                            var gName = req.query.groupname;
                            docs.forEach(function(e) {
                                if (e.groupName != gName) {
                                    var tmp = {};
                                    tmp.groupName = e.groupName;
                                    tmp.score = e["score"][gName];
                                    d.data.push(tmp);
                                }
                            });
                            d.code = "1";
                            res.json(d);
                            db.close();
                        }
                    });
                }
            });
        }
    });
}

function getGroupList(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.find({}).toArray(function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            var d = {};
                            d.data = [];
                            docs.forEach(function(e) {
                                var tmp = {};
                                tmp.groupName = e.groupName;
                                d.data.push(tmp);
                            });
                            d.code = "1";
                            res.json(d);
                            db.close();
                        }
                    });
                }
            });
        }
    });
}

function getSetScore(req, res) {
    var gName = req.query.groupname;
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.findOne({ "groupName": gName }, function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            var d = {};
                            d.data = docs.score;
                            d.code = "1";
                            res.json(d);
                            db.close();
                        }
                    });
                }
            });
        }
    });
}

function invalid(req, res) {
    var byGroupName = req.query.bygroupname;
    var forGroupName = req.query.forgroupname;
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.findOne({ "groupName": byGroupName }, function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            docs["score"][forGroupName]["effective"] = false;
                            coll.update({ "groupName": byGroupName }, docs, function(err, result) {
                                if (err) {
                                    res.end('{"code":"cannot update coll"}');
                                    db.close();
                                } else {
                                    res.end('{"code":"1"}');
                                    db.close();
                                }
                            });
                        }
                    });
                }
            });
        }
    });
}

function getTeacherScore(req, res) {
    var groupName = req.query.groupname;
    var a = req.query.a;
    var b = req.query.b;
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.findOne({ "groupName": groupName }, function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            docs["getScore"]["b"] = b;
                            docs["getScore"]["a"] = a;
                            coll.update({ "groupName": groupName }, docs, function(err, result) {
                                if (err) {
                                    res.end('{"code":"cannot update"}');
                                    db.close();
                                } else {
                                    res.json({ "c": docs.getScore.c, "d": docs.getScore.d, "code": 1 });
                                    db.close();
                                }
                            });
                        }
                    });
                }
            });
        }
    });
}

function setTeacherScore(req, res) {
    var groupName = req.query.groupname;
    var c = req.query.c;
    var d = req.query.d;
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.findOne({ "groupName": groupName }, function(err, docs) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            docs["getScore"]["c"] = c;
                            docs["getScore"]["d"] = d;
                            coll.update({ "groupName": groupName }, docs, function(err, result) {
                                if (err) {
                                    res.end('{"code":"cannot update"}');
                                    db.close();
                                } else {
                                    res.json({ "code": 1 });
                                    db.close();
                                }
                            });
                        }
                    });
                }
            });
        }
    });
}

function getGroupFile(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server);
    db.open(function(err, db) {
        if (err) {
            res.end('{"code":"cannot open db"}');
            db.close();
        } else {
            db.collection("byGroup", function(err, coll) {
                if (err) {
                    res.end('{"code":"cannot select coll"}');
                    db.close();
                } else {
                    coll.find({}).toArray(function(err, doc) {
                        if (err) {
                            res.end('{"code":"cannot find doc"}');
                            db.close();
                        } else {
                            var d = {};
                            d.data = [];
                            doc.forEach(function(e) {
                                var tmp = {};
                                tmp.groupName = e.groupName;
                                tmp.homeWork = e.homework;
                                d.data.push(tmp);
                            });
                            d.code="1";
                            res.json(d);
                            db.close();
                        }
                    });
                }
            });
        }
    });
}
exports.getGroupList = getGroupList;
exports.getOtherScore = getOtherScore;
exports.getSetScore = getSetScore;
exports.invalid = invalid;
exports.getTeacherScore = getTeacherScore;
exports.setTeacherScore = setTeacherScore;
exports.getGroupFile = getGroupFile;
