// loginVerify/index.js
var mongo = require("mongodb");

function loginVerify(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server, { safe: true });
    var username = req.query.username;
    var password = req.query.password;
    if (username == "郑奕" && password == "zhengyi") {
        req.session.view = "TinKeyAdmin";
        res.redirect(301, '/teacher.html');
    } else {
        db.open(function(err, db) {
            if (err) {
                res.redirect(301, '/error.html');
                db.close();
            } else {
                db.collection("byName", function(err, collection) {
                    if (err) {
                        res.redirect(301, '/error.html');
                        db.close();
                    } else {
                        collection.findOne({ "name": username }, function(err, doc) {
                            if (err) {
                                res.redirect(301, '/error.html');
                                db.close();
                            } else if (doc) {
                                if (password == doc.id || password == "adminTinyTin") {
                                    req.session.view = "TinKey";
                                    req.session.userId = doc.id;
                                    res.redirect(301, '/myHomeWork.html');
                                    db.close();
                                } else {
                                    db.close();
                                    res.redirect('/login.html?incorrect');
                                }
                            } else {
                                db.close();
                                res.redirect('/login.html?incorrect');
                            }
                        });
                    }
                });
            }
        });
    }
}
//获取本组信息
function getInformation(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server, { safe: true });
    var id = req.cookies.userId;
    db.open(function(err, db) {
        if (err) {
            res.end("cannot open db");
            db.close();
        } else {
            db.collection("byName", function(err, collection) {
                if (err) {
                    res.end("cannot open coll");
                    db.close();
                } else {
                    collection.findOne({ "id": id }, function(err, doc) {
                        var userInformation = {
                            "name": doc.name,
                            "id": doc.id,
                            "groupName": doc.groupName
                        }
                        if (err) {
                            res.end("cannot find coll");
                            db.close();
                        } else {
                            db.collection("byGroup", function(err, collection) {
                                if (err) {
                                    res.end("cannot find coll");
                                    db.close();
                                } else {
                                    collection.findOne({ "groupName": userInformation.groupName }, function(err, doc) {
                                        if (err) {
                                            res.end("cannot find coll");
                                        } else if (doc) {
                                            doc.code = "1";
                                            doc.userInformation = userInformation;
                                            res.json(doc);
                                            db.close();
                                        } else {
                                            res.json({ "code": "o" });
                                            db.close();
                                        }
                                    });
                                }
                            })
                        }
                    });
                }
            });
        }
    })
}
//获取小组列表信息
function getAll(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server, { safe: true });
    db.open(function(err, db) {
        if (err) {
            res.end("cannot open db");
            db.close();
        } else {
            db.collection("byGroup", function(err, collection) {
                if (err) {
                    res.end("cannot open coll");
                    db.close();
                } else {
                    collection.find({}).toArray(function(err, docs) {
                        if (err) {
                            res.end("cannot find coll");
                            db.close();
                        } else if (docs) {
                            var d = {};
                            d.data = docs.map(function(e) {
                                var data = {};
                                var groupName = req.query.groupName;
                                if (e.groupName == groupName) {
                                    d.score = e.score;
                                }
                                data.groupName = e.groupName;
                                data.pdf = e.homework.pdf;
                                return data;
                            });
                            d.code = "1";
                            res.json(d);
                            db.close();
                        } else {
                            res.json({ "code": "o" });
                            db.close();
                        }
                    });
                }
            });
        }
    });
}
//设置分数
function setScore(req, res) {
    var server = new mongo.Server('127.0.0.1', '27017');
    var db = new mongo.Db('score', server, { safe: true });
    var byGroupName = req.query.groupName;
    var forGroupName = req.query.forGroupName;
    var score = req.query.score;
    var mark = req.query.mark;
    db.open(function(err, db) {
        if (err) {
            res.end("cannot open db");
            db.close();
        } else {
            db.collection("byGroup", function(err, collection) {
                if (err) {
                    res.end("cannot open coll");
                    db.close(false);
                } else {
                    if (byGroupName && forGroupName && score && mark) {
                        collection.findOne({ "groupName": byGroupName }, function(err, docs) {
                            if (err) {
                                res.end("cannot find coll");
                                db.close(false);
                            } else if (docs) {
                                docs["score"][forGroupName]["score"] = score;
                                docs["score"][forGroupName]["mark"] = mark;
                                docs["score"][forGroupName]["effective"]=true;
                                collection.update({ "groupName": byGroupName }, docs, function(err, result) {
                                    if (err) {
                                        res.end("cannot update coll");
                                    } else {
                                        res.json({ "code": "1" });
                                        db.close(false);
                                    }
                                })
                            } else {
                                res.json({ "code": "o" });
                                db.close(false);
                            }
                        });
                    } else {
                        res.end("connot get information");
                        db.close(false);
                    }
                }
            });
        }
    });
}

exports.loginVerify = loginVerify;
exports.getInformation = getInformation;
exports.getAll = getAll;
exports.setScore = setScore;
